tests/Tagged/TaggedProviderTest.php

```diff
@@ -0,0 +1,121 @@
+<?php
+declare(strict_types=1);
+
+namespace Fig\EventDispatcher\Tagged;
+
+use Fig\EventDispatcher\ParameterDeriverTrait;
+use Fig\EventDispatcher\TaggedProviderTrait;
+use Fig\EventDispatcher\CollectingEvent;
+use PHPUnit\Framework\TestCase;
+use Psr\EventDispatcher\ListenerProviderInterface;
+
+class TaggedProviderTest extends TestCase
+{
+
+    /** @var ListenerProviderInterface */
+    protected $provider;
+
+    public function setUp()
+    {
+        parent::setUp();
+
+        $this->provider = new class implements ListenerProviderInterface {
+            use TaggedProviderTrait;
+            use ParameterDeriverTrait;
+
+            /** @var array */
+            protected $listeners = [];
+
+            /** @var array */
+            protected $all = [];
+
+
+            protected function eventType(): string
+            {
+                return WorkflowEventInterface::class;
+            }
+
+            protected function tagMethod(): string
+            {
+                return 'workflowName';
+            }
+
+            public function addListener(callable $listener, string $tagName = '', string $type = null) : void
+            {
+                $type = $type ?? $this->getParameterType($listener);
+
+                if ($tagName) {
+                    $this->listeners[$tagName][$type][] = $listener;
+                }
+                else {
```

Comment:
PSR-1/2 have you put the `else` on the same line as the previous closing brace.

Also, this _could_  become:

```php
if ($tagName) {
    $this->listeners[$tagName][$type][] = $listener;
    return;
}
$this->all[$type][] = $listener;
```

which would obviate the need for the `else` entirely.